src/constbuffer_array_splitter.c

```diff
@@ -208,3 +208,248 @@ CONSTBUFFER_ARRAY_HANDLE constbuffer_array_splitter_split(CONSTBUFFER_ARRAY_HAND
+                                            buffers, start_buffer_index, 1, start_buffer_offset, max_buffer_size);
+                                        CONSTBUFFER_DecRef(curr_buffer);
+                                        break;
+                                    }
+                                    else
+                                    {
+                                        temp->arr[index] = arr1;
+                                        buffer_array_count++;
+                                        index++;
+                                        current_buffer_remaining_size -= max_buffer_size;
+                                        start_buffer_offset += max_buffer_size;
+                                    }
+                                }
+                                //reset for another sub array
+                                if (current_buffer_remaining_size == 0)
+                                {
+                                    start_buffer_index = i + 1;
+                                    start_buffer_offset = 0;
+                                    current_buffer_count = 0;
+                                    current_buffer_size = 0;
+                                }
+                                else
+                                {
+                                    /* Codes_SRS_CONSTBUFFER_ARRAY_SPLITTER_07_014: [ If current buffer is the last buffer in the original constbuffer_array, constbuffer_array_splitter_split_to_array_of_array shall store the sub - tarray with size smaller than max_buffer_size to result. ]*/
+                                    if (i == buffer_count - 1)
+                                    {
+                                        CONSTBUFFER_ARRAY_HANDLE arr2 = constbuffer_array_create_from_buffer_offset_and_count(buffers, start_buffer_index, 1, start_buffer_offset, current_buffer_remaining_size);
+                                        if (arr2 == NULL)
+                                        {
+                                            /* Codes_SRS_CONSTBUFFER_ARRAY_SPLITTER_07_004: [ If there are any other failures then constbuffer_array_splitter_split_to_array_of_array shall fail and return NULL. ]*/
+                                            failed = true;
+                                            LogError("constbuffer_array_create_from_buffer_offset_and_count failed buffers=%p, uint32_t start_buffer_index=%" PRIu32 ", uint32_t buffer_count=%" PRIu32 ", uint32_t start_buffer_offset=%" PRIu32 ", uint32_t end_buffer_size=%" PRIu32,
+                                                buffers, start_buffer_index, 1, start_buffer_offset, max_buffer_size);
+                                            CONSTBUFFER_DecRef(curr_buffer);
+                                            break;
+                                        }
```

Comment:
I don't quite like the fact that we break and dec ref on each failure path.
It makes the code brittle. Imagine if someone changes the code and forgets to do the DecRef on an error path. That is a bug. Which will likely get caught by unit tests if we have coverage. We generally do have coverage, but point is that we can write the code slightly better to make it less brittle.

We could break if failed is set to true between lines 433 and 434. Then we would not need to do DecRef on each failure path

Reply:
done